ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchConfigurationHelper.java

```diff
@@ -229,40 +231,29 @@ private static String computeOSGiBundles(String bundleList, Map<String, List<IPl
 		if (bundles.get(IPDEBuildConstants.BUNDLE_SIMPLE_CONFIGURATOR) != null)
 			return "org.eclipse.equinox.simpleconfigurator@1:start"; //$NON-NLS-1$
 
-		StringBuilder buffer = new StringBuilder();
-		Set<String> initialBundleSet = new HashSet<>();
-		StringTokenizer tokenizer = new StringTokenizer(bundleList, ","); //$NON-NLS-1$
+		Map<IPluginModelBase, String> allBundles = new HashMap<>(bundlesWithStartLevels);
+		allBundles.keySet().removeIf(model -> IPDEBuildConstants.BUNDLE_OSGI.equals(model.getPluginBase().getId())); // write out all bundles in osgi.bundles - bug 170772
+		// then override with resolved initialBundleList content
+		StringTokenizer tokenizer = new StringTokenizer(initialBundleList, ","); //$NON-NLS-1$
 		while (tokenizer.hasMoreTokens()) {
 			String token = tokenizer.nextToken();
 			int index = token.indexOf('@');
 			String id = index != -1 ? token.substring(0, index) : token;
-			if (bundles.containsKey(id)) {
-				if (buffer.length() > 0)
-					buffer.append(',');
-				buffer.append(id);
-				if (index != -1 && index < token.length() - 1)
-					buffer.append(token.substring(index));
-				initialBundleSet.add(id);
-			}
-		}
-
-		// write out all bundles in osgi.bundles - bug 170772
-		initialBundleSet.add(IPDEBuildConstants.BUNDLE_OSGI);
-		for (Entry<IPluginModelBase, String> entry : bundlesWithStartLevels.entrySet()) {
-			IPluginModelBase model = entry.getKey();
-			String id = model.getPluginBase().getId();
-			if (!initialBundleSet.contains(id)) {
-				if (buffer.length() > 0)
-					buffer.append(',');
-
-				String slinfo = entry.getValue();
-				buffer.append(id);
-				buffer.append('@');
-				buffer.append(slinfo);
+			IPluginModelBase model = getLatestModel(id, bundles);
+			if (model != null) {
+				// when present, startInfo from intialBundlesList overrides the one from the config
+				String startInfo = index + 1 < token.length() ? token.substring(index + 1) : allBundles.get(model);
```

Comment:
I think the start-info from the intialBundlesList should only overwrite an exsting start-level or autostart value if that value is empty or `"default"`. A value specified in the launch config should take precedence over a default value.

Reply:
The current (before this patch) behavior seems to be that if the bundle comes from initialBundlesList, then other startInfo settings are skipped ( https://github.com/eclipse-pde/eclipse.pde/blob/3f9f33643122a8109460e88362593d67fde42ed5/ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchConfigurationHelper.java#L254 excludes processing of bundles coming from initialBundlesList ); so basically it means that initialBundlesList takes higher priority over other settings, ie it can override them.
So I believe this change keeps the current state. Or am I missing 1 case?

Reply:
Indeed. I didn't see another case. Then lets keep the behavior as it is in this PR (which is also simpler).
Nevertheless I think it would be better to prefer the value from the launch config. But this can be done in a follow up.